vtex/faststore
# Feat: VTEX API Schema extensions

## What's the purpose of this pull request?

- [x] Creates the basic scaffold of API extensions types and resolvers based on API Extensions RFC.
- [x] Export Root object fields so that we can use it on extension resolvers. 
- [x] Aims to merge the extensions GraphQL with the native. 
- [x] ~~Adds a basic example.~~
- [x] Adds [API Extension Docs](https://faststore-site-5hu0dq0lh-faststore.vercel.app/docs/api-extensions) with Overview and VTEX extension API section: 


## How it works?

As stores tend to grow, so does the possibility of consuming data from VTEX APIs that are not exposed by the FastStore API. This PR enables to use extending the FastStore API by using the `packages/core/src/customizations/graphql/vtex` folder.

you can also follow [this doc](https://faststore-site-5hu0dq0lh-faststore.vercel.app/docs/api-extensions)

This folder is a mirror from the one from the `starter.store` src folder that will be done in another PR.

## How to test it?
Follow [this doc](https://faststore-site-5hu0dq0lh-faststore.vercel.app/docs/api-extensions) steps.

And then:
1. add the `customData` field in the `ProductDetailsFragment_product` from the `packages/core/src/components/sections/ProductDetails/ProductDetails.tsx`.
2. run `yarn dev` in the core package.
3. run `yarn generate` in the core package (with `yarn dev` running).
4. Get and use the `customData` as you want in the `packages/core/src/components/sections/ProductDetails/ProductDetails.tsx`. 
Example:
<img width="434" alt="Screenshot 2023-07-25 at 20 31 44" src="https://github.com/vtex/faststore/assets/11325562/e5d0c0ac-6a09-44a3-a2b7-f9e2dcd09882">

<br><br>

You could also test the `typeDefs` fallback option by renaming the `typeDefs` folder:
1. the name `typeDefs` should work.
2. the name `typedefs` should work as fallback. 
3. Other names should throw an error. 

### Starters Deploy Preview

TBD

## References
https://the-guild.dev/graphql/tools/docs/schema-merging#file-loading
